### PR TITLE
2.x: Expand the Getting Started

### PR DESCRIPTION
**Pull request by akarnokd**
This PR expands the Readme.md with further examples and details, plus adds headers so that (former and new) sections can be easily linked in answers on issue lists or StackOverflow.

(Note that it was written offline without proper markdown rendering so I may update the PR a couple of times to get the render mistakes fixed).

### CODE REVIEW

**Comment by akarnokd** [comment]
Haven't encountered any rendering problems.

File: README.md [hunk 1]
```diff
@@ -66,13 +73,93 @@ Flowable.just("Hello world")
...
+  - [`io.reactivex.Maybe`](http://reactivex.io/RxJava/2.x/javadoc/io/reactivex/Maybe.html): a flow with no items, exactly one item or an error.
+
+### Some terminology
+
+#### Upstream, downstream
+
+The dataflows in RxJava consist of a source, zero or more intermediate steps followed by a data consumer or combinator step (where the step is responsible to consume the dataflow by some means):
+
+```java
+
```
Comment (vanniktech) [inline README.md]: are those new lines necessary?

**Review by vanniktech — APPROVED** [review]

File: README.md (hunk 1 above)
Comment (akarnokd) [reply README.md]: Those seem to be stripped from the final render:

image

Comment (akarnokd) [reply README.md]: In this case though, there was a line there which got accidentally deleted.

**Review by vanniktech — APPROVED** [review]

File: README.md [hunk 2]
```diff
@@ -33,12 +33,19 @@ See the differences between version 1.x and 2.x in the wiki article [What's diff
...
 
+### Setting up the dependency
+
 The first step is to include RxJava 2 into your project, for example, as a Gradle compile dependency:
 
 ```groovy
 compile "io.reactivex.rxjava2:rxjava:2.x.y"
 ```
 
+(Please replace `x` and `y` with the latest version numbers: [![Maven Central](https://maven-badges.herokuapp.com/maven-central/io.reactivex.rxjava2/rxjava/badge.svg)](https://maven-badges.herokuapp.com/maven-central/io.reactivex.rxjava2/rxjava)
```
Comment (artem-zinnatullin) [inline README.md:44]: Badge image is not displaying, I think `[` before `!` is not needed

https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#images

Comment (artem-zinnatullin) [reply README.md:44]: Ah, you want to make it clickable, then something wrong with syntax or my browser (Firefox 59.0.1)

Comment (akarnokd) [reply README.md:44]: It is the same markdown as in the beginning of the readme file where the top badges are located. Sometimes the maven badge simply fails to load. I'm assuming this is due to throttling at the host service where maven-badges lives.

File: README.md [hunk 3]
```diff
@@ -66,13 +73,93 @@ Flowable.just("Hello world")
...
-  - [`io.reactivex.Observable`](http://reactivex.io/RxJava/2.x/javadoc/io/reactivex/Observable.html): 0..N flows, no backpressure
-  - [`io.reactivex.Single`](http://reactivex.io/RxJava/2.x/javadoc/io/reactivex/Single.html): a flow of exactly 1 item or an error
-  - [`io.reactivex.Completable`](http://reactivex.io/RxJava/2.x/javadoc/io/reactivex/Completable.html): a flow without items but only a completion or error signal
-  - [`io.reactivex.Maybe`](http://reactivex.io/RxJava/2.x/javadoc/io/reactivex/Maybe.html): a flow with no items, exactly one item or an error
+  - [`io.reactivex.Observable`](http://reactivex.io/RxJava/2.x/javadoc/io/reactivex/Observable.html): 0..N flows, no backpressure,
+  - [`io.reactivex.Single`](http://reactivex.io/RxJava/2.x/javadoc/io/reactivex/Single.html): a flow of exactly 1 item or an error,
+  - [`io.reactivex.Completable`](http://reactivex.io/RxJava/2.x/javadoc/io/reactivex/Completable.html): a flow without items but only a completion or error signal,
+  - [`io.reactivex.Maybe`](http://reactivex.io/RxJava/2.x/javadoc/io/reactivex/Maybe.html): a flow with no items, exactly one item or an error.
+
+### Some terminology
```
Comment (artem-zinnatullin) [inline README.md:86]: nit: "Some" → "Common"?

File: README.md [hunk 4]
```diff
@@ -66,13 +73,93 @@ Flowable.just("Hello world")
...
+
+#### Upstream, downstream
+
+The dataflows in RxJava consist of a source, zero or more intermediate steps followed by a data consumer or combinator step (where the step is responsible to consume the dataflow by some means):
```
Comment (artem-zinnatullin) [inline README.md:90]: "combinator" → "operator"?

I see few other usages of "combinator" https://github.com/ReactiveX/RxJava/search?q=combinator&type=Code&utf8=%E2%9C%93 but idk, doesn't look like common term in RxJava community

File: README.md [hunk 5]
```diff
@@ -66,13 +73,93 @@ Flowable.just("Hello world")
...
+
+```java
+source.operator1().operator2().operator3().subscribe(consumer);
```
Comment (artem-zinnatullin) [inline README.md:93]: Here we see "operator" instead of "combinator", so something should be changed for consistency I guess

File: README.md [hunk 6]
```diff
@@ -66,13 +73,93 @@ Flowable.just("Hello world")
...
+
+```java
+source
+  .operator1()
+  .operator2()
+  .operator3()
+  .subscribe(consumer)
+```
+
+#### Objects in motion
```
Comment (artem-zinnatullin) [inline README.md:108]: Never heard "Objects in motion" used to describe RxJava streams

https://github.com/reactive-streams/reactive-streams-jvm#glossary

Can't come up with good alternative atm, something along with "Dataflow" could work

File: README.md [hunk 7]
```diff
@@ -66,13 +73,93 @@ Flowable.just("Hello world")
...
+
+#### Subscription time
+
+This is a temporary state when `subscribe()` is called on a flow that establishes the chain of processing steps internally:
+
+```java
+flow.subscribe(System.out::println)
+````
+
+This is when the **subscription side-effects** are triggered (see `doOnSubscribe`). Some sources block or start emitting items right away in this state.
```
Comment (artem-zinnatullin) [inline README.md:139]: Maybe reword to something like: 

>"Typically in this state source starts synchronous or asynchronous work to prepare data to be emitted."

?

File: README.md [hunk 8]
```diff
@@ -66,13 +73,93 @@ Flowable.just("Hello world")
...
+
+#### Runtime
```
Comment (artem-zinnatullin) [inline README.md:141]: I know it's subjective, but `Runtime` reads as noun, an object here rather than set of states in time.

Maybe change to "Run-time"?

"“Runtime”, “run time”, and “run-time”": https://english.stackexchange.com/a/197688

**Review by artem-zinnatullin — COMMENTED** [review]
Great rewrite!

Sorry can't finish it right now (don't want to block merge though), posting partial review

File: README.md (hunk 3 above)
Comment (akarnokd) [reply README.md:86]: These are not all the terminology but those which appear later in the text and I felt I have to define them upfront.

File: README.md (hunk 4 above)
Comment (akarnokd) [reply README.md:90]: https://github.com/ReactiveX/RxJava/wiki/Combining-Observables

File: README.md (hunk 5 above)
Comment (akarnokd) [reply README.md:93]: Here operators are the steps, the combinator refers to `flatMap`.

File: README.md (hunk 6 above)
Comment (akarnokd) [reply README.md:108]: I had to encompass all the terms below without repeating them in the title. This felt like a nice encompassing term.

File: README.md (hunk 7 above)
Comment (akarnokd) [reply README.md:139]: I think referring to "block" and "emitting" is more appropriate in a getting started guide.

File: README.md (hunk 8 above)
Comment (akarnokd) [reply README.md:141]: I think we need a native English speaker for this.

File: README.md (hunk 6 above)
Comment (artem-zinnatullin) [reply README.md:108]: yeah I understand, up2u of course

But just in case, headers in markdown can be clicked and used in urls, so changes in future will break existing urls

That's why I decided to comment it even though I don't have a good replacement option atm

Comment (akarnokd) [reply README.md:108]: I'll try to remember and add hidden anchors `<a name="objects-in-motion"></a>` if the time comes.
